Bigtable/src/Instance.php

```diff
@@ -0,0 +1,271 @@
+    }
+
+    /**
+     * Check if the instance exists.
+     *
+     * This method requires a service call.
+     *
+     *
+     * @param array $options [optional] Configuration options.
+     */
+    public function exists(array $options = [])
+    {
+        throw new \BadMethodCallException('This method is not implemented yet');
+    }
+
+    /**
+     * Fetch a fresh representation of the instance from the service.
+     *
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/bigtable/docs/reference/admin/rpc/google.bigtable.admin.v2#google.bigtable.admin.v2.GetInstanceRequest GetInstanceRequest
+     * @see https://cloud.google.com/bigtable/docs/reference/admin/rpc/google.bigtable.admin.v2#instance Instance
+     * @codingStandardsIgnoreEnd
+     *
+     * @param array $options [optional] Configuration options.
+     */
+    public function reload(array $options = [])
+    {
+        throw new \BadMethodCallException('This method is not implemented yet');
+    }
+
+    /**
+     * Return the instance state.
+     *
+     * When instances are created or updated, they may take some time before
+     * they are ready for use. This method allows for checking whether an
+     * instance is ready.
+     *
+     *
+     * @param array $options [optional] Configuration options.
+     */
+    public function state(array $options = [])
+    {
+        throw new \BadMethodCallException('This method is not implemented yet');
+    }
+
+    /**
+     * Create a new instance.
+     *
+     * Example:
+     * ```
+     * use Google\Cloud\Bigtable\BigtableClient;
+     * $bigtable = new BigtableClient();
+     * $instanceType = Google\Cloud\Bigtable\Instance::INSTANCE_TYPE_DEVELOPMENT;
+     * $operation = $instance->create(
+     *     'instanceId',
+     *     $instanceType,
+     *     ['foo' => 'bar'],
+     *     [$bigtable->clusterMetadata('my-cluster', 'us-east1-b')]
+     * );
+     * ```
+     *
+     * @codingStandardsIgnoreStart
+     * @see https://cloud.google.com/bigtable/docs/reference/admin/rpc/google.bigtable.admin.v2#CreateInstanceRequest CreateInstanceRequest
+     *
+     * @param string $displayName **Defaults to** the value of $instanceId.
+     * @param int $type Possible values are represented by the following constants: 
+     *            `Google\Cloud\Bigtable\Instance::INSTANCE_TYPE_PRODUCTION`, 
+     *            `Google\Cloud\Bigtable\Instance::INSTANCE_TYPE_DEVELOPMENT` and
+     *            `Google\Cloud\Bigtable\Instance::INSTANCE_TYPE_UNSPECIFIED`.
+     *            **Defaults to** using `Google\Cloud\Bigtable\Instance::INSTANCE_TYPE_UNSPECIFIED`.
+     * @param array $labels as key/value pair ['foo' => 'bar']. For more information, see
+     *        [Using labels to organize Google Cloud Platform resources](https://cloudplatform.googleblog.com/2015/10/using-labels-to-organize-Google-Cloud-Platform-resources.html).
+     * @param array $clustersMetadata [
```

Comment:
This is confusing because the variable itself is $clustersMetadata and it's an array.
Its elements are also a result of a method with similar name "clusterMetadata".

Can we make this easier to understand?

Reply:
This is definitely hard to understand.  Do you have any suggestions?  I'm happy to VC with you about the inherent complexities.

Reply:
From what [I've studied](https://cloud.google.com/bigtable/docs/instances-clusters-nodes), an instance can have at most two clusters, and each cluster can have at most 3 nodes. (Please correct me if I'm wrong).

So, what we pass to this `create()` here should be actually "a list of clusters", as all information returned by `$bigtable->clusterMetadata()` actually belongs to a cluster.
I'm not sure what @dwsupplee  and @jdpedrie  have discussed last week. It looks like we decided to use an array to represent a cluster instead of a class?

The quick and easiest way to fix this would be to rename methods and params here.
What about renaming `clusterMetadata()` to just a `cluster()` or `clusterInfo()`?
Ultimately, what we want to create here is to represent a cluster, so just `cluster()` should be fine.
The term metadata seems confusing because it can be interpreted as "metadata of a cluster" or "metadata of instance (where this cluster is passed)".
Then we can rename the param name from `$clustersMetadata` to `$clusters` or `$clusterList`.
This should already clear things up.

BTW, I think this is a good example showing that creating a class to represent a cluster will really help.
If we were to create a `Cluster` class, the signature of this `create()` method would look like this:

```php
/**
 * (some stuff)
 * @param Cluster[] $clusters a list of clusters that this instance will create.
 *                  An instance can have up to 2 clusters.
 * (some stuff)
 */
public function create($displayName, $type, $labels, array $clusters)
```
This should probably be much cleaner and doesn't require long PHPDoc. We can also put more info in a Cluster class, if needed.
You can also have more symmetric API designs too as you'll have a `$bigtableClient->instance()` to create an `Instance` and `$bigtableClient->cluster()` to create a `Cluster`.

Finally, as a small concern, it looks like `clusterMetadata()` (or `cluster()` if we rename it) doesn't depend on `BigtableClient` properties.
Do we have any reasons why we don't make it a static method? Is it just a convention or to keep consistency with `instance()`?

Reply:
Naming is difficult, so thank you for your thoughts on how to rename methods.

I have a few observations:

1. There's a temporary limitation on the number of clusters.  Eventually the service will allow more than two.

2. A `Cluster` object makes sense for manipulating an existing `Cluster`. Specifically, `updateCluster` and `deleteCluster` operations make sense on a `Cluster` object.  `createInstance` or `createCluster` can happen via a `Cluster` object, but a `clusterMetadata` object (or an array) are simpler to understand, IMHO.

3. A `Cluster` object is a "child" of an instance.  Instances have a pattern like `projects/{projectId}/instances/{instanceId}`.  Clusters have a pattern like `projects/{projectId}/instances/{instanceId}/clusters/{clusterId}`.  it's more fluid to create `Cluster` objects via `$bigtableClient->instance('instance-id')->cluster('cluster-id')`; the alternative would be `$bigtableClient->instance('instance-id', 'cluster-id')`.  Either way, creating a Cluster object doesn't make sense, or at least is confusing, for the `createInstance` method.

Let's set up some time to discuss this offline.

Reply:
Naming is difficult, so thank you for your thoughts on how to rename methods.

I have a few observations:

1. There's a temporary limitation on the number of clusters.  Eventually the service will allow more than two.

2. A `Cluster` object makes sense for manipulating an existing `Cluster`. Specifically, `updateCluster` and `deleteCluster` operations make sense on a `Cluster` object.  `createInstance` or `createCluster` can happen via a `Cluster` object, but a `clusterMetadata` object (or an array) are simpler to understand, IMHO.

3. A `Cluster` object is a "child" of an instance.  Instances have a pattern like `projects/{projectId}/instances/{instanceId}`.  Clusters have a pattern like `projects/{projectId}/instances/{instanceId}/clusters/{clusterId}`.  it's more fluid to create `Cluster` objects via `$bigtableClient->instance('instance-id')->cluster('cluster-id')`; the alternative would be `$bigtableClient->instance('instance-id', 'cluster-id')`.  Either way, creating a Cluster object doesn't make sense, or at least is confusing, for the `createInstance` method.

Let's set up some time to discuss this offline.

Reply:
> Then we can rename the param name from $clustersMetadata to $clusters or $clusterList.

+1000. I prefer `$clusters` personally.

Reply:
Solomon, thanks for more insight. :)
Sorry for the confusion. I've just realized that we also have a generated class named Cluster too.

For 2), I don't feel strongly against if you really need a clusterMetadata object. It seems like a one-time piece of information for creating an instance, right? In that case, 
* We can retain the name `clusterMetadata` (though I prefer `clusterInfo` as it's an "info" about clusters of the "instance" we're going to create), but we need to change the param name to "clusterMetadataList" for example. It's quite tricky as clusterMetadata is already plural and we can't do "clusters" as I originally suggested.

For 3), I totally agree on that (I observed that too).
I just thought that we need to stick to the design pattern that everything is created from `BigtableClient`. To me, it makes much much more sense to create cluster via `Instance` instead (or at least via `instance('instance-id', 'cluster-id')`).